[WIP]Milking Animals

AlecWhite opened the pull request:
I don't know how a PR should be described/explained/documented when it touches the source, and this is the first time I edit a program in C++ that wasn't exclusively part of an exercise from learncpp.com, so I apologize for any mistakes and any help is welcome.

This PR introduce the MILKABLE flag with the necessary code to milk animals.

What this PR adds:
-Added MILKABLE flag for monsters.
-Milking gives 4 units of milk and is hardcoded.
-Milking has a 24 hours cool-down and is hardcoded.
-Milking message saying what you are milking and is hardcoded right now to say "cow".

Started this once I made sure that I couldn't mod it with the current Lua voodoo. 
And because it is a PR to merged and not just a random mod I make, I don't know what way of balancing it would be best, so I just added the bare minimum to it to make it workable and not too OP.
For when it was for a mod, I thought of needing a "tamed" or "genetically enhanced" cow that only spawns in a new "diary farm". Said cow would not attack the player or flee from her, and in general be almost brain dead thanks to the "genetically enhancing", which would make you have to protect your cows. 

But now that it is done in the source, I guess it could need one of the above with the extra of having a calf stay on X tiles around the cow, and being fed. Or something else like needing survival skill to be able to milk. I mean now that this is in the source and not just a mod the possibilities are endless. 
But then of course I'm a noob at C++ and this PR took me the better part of 10 hours between compile errors, looking what I could use on the already existing code and trying to implement it with my limited skills. Luckily Wulfe813 and other people from the IRC channel helped me.

AlecWhite commented:
All checks have failed and I don't know even how to read them. 
I could use some help please.

AMurkin commented:
Error from travis:
```
src/iexamine.cpp:2785:6: error: unused parameter ‘examp’ [-Werror=unused-parameter]
 void iexamine::milk_source(const tripoint &examp, monster *source_mon)
      ^
```

AlecWhite commented:
Thanks, now there's only one check left.

Leland commented:
Looks like you have astyle issues. Try running `make astyle-all` and see how the code you're committing here changes. You may need to install astyle if you don't already have it.

There are quite a few that should be immediately obvious if you go through the diff on this PR: https://github.com/CleverRaven/Cataclysm-DDA/pull/21042/files.

Leland commented on src/game.h:
```diff
@@ -644,7 +645,8 @@ class game
                             const tripoint *source_pos = nullptr,
-                            const vehicle *source_veh = nullptr );
+                            const vehicle *source_veh = nullptr,
+							const monster *source_mon = nullptr );
```
Needs an extra space

Leland commented on src/creature.h:
```diff
@@ -543,6 +543,9 @@ class Creature
         virtual const std::string &symbol() const = 0;
         virtual bool is_symbol_highlighted() const;
+		// Turn since creature was milked for last time
+		int turn_next_milking = 0;
```
Needs an extra space

Leland commented on src/iexamine.cpp:
```diff
@@ -2780,6 +2782,20 @@ void iexamine::water_source(player &p, const tripoint &examp)
+	}else {
+		add_msg("The cow's udders run dry");
+	}
+	
```
Unneeded line

AlecWhite commented:
Installed Astyle for Visual Studio, fed the commands in CODE_STYLE  and edited the bits I added or modify.
If I run the Astyle on the whole file it changes almost all the lines. So I had just to feed the bits I changed.

Leland commented:
@AlecWhite try taking a look at the diff for this PR. Should be pretty obvious where the astyle errors are. https://github.com/CleverRaven/Cataclysm-DDA/pull/21042/files

AlecWhite commented:
@Leland Where and how do I run make astyle-all to see the issues?

Leland commented:
@AlecWhite running `make astyle-all` from the CLI will make a massive amount of changes, but just look specifically at your blocks of code to see how they've changed. Then you can just copy your affected blocks, `git checkout` the files to revert them, then paste the fixed code back.

Leland commented on src/mtype.h:
```diff
@@ -157,6 +157,7 @@ enum m_flag : int {
     MF_AVOID_DANGER_2,      // This monster will path around most dangers instead of through them.
     MF_PRIORITIZE_TARGETS,  // This monster will prioritize targets depending on their danger levels
     MF_NOT_HALLU,           // Monsters that will NOT appear when player's producing hallucinations
+	MF_MILKABLE,            // This monster is milkable, applies only to cows.
```
Needs to be indented

Leland commented on src/monstergenerator.cpp:
```diff
@@ -453,6 +453,7 @@ void MonsterGenerator::init_flags()
     flag_map["PATH_AVOID_DANGER_2"] = MF_AVOID_DANGER_2;
     flag_map["PRIORITIZE_TARGETS"] = MF_PRIORITIZE_TARGETS;
     flag_map["NOT_HALLUCINATION"] = MF_NOT_HALLU;
+	flag_map["MILKABLE"] = MF_MILKABLE;
```
Needs to be indented

Leland commented on src/iexamine.cpp:
```diff
@@ -2780,6 +2782,19 @@ void iexamine::water_source(player &p, const tripoint &examp)
+    } else {
+        add_msg( "The cow's udders run dry" );
+    }
+
```
Kill this extra line

Leland commented on src/game.h:
```diff
@@ -635,6 +635,7 @@ class game
          * @param radius Radius to look for liquid around pos
          * @param source_pos The source of the liquid when it's from the map.
          * @param source_veh The vehicle that currently contains the liquid in its tank.
+		 * @param source_mon The source monster that is milked.
```
Needs another space

Leland commented on src/game.cpp:
```diff
@@ -9761,7 +9767,8 @@ extern void serialize_liquid_target( player_activity &act, const tripoint &pos )
                           const tripoint * const source_pos,
-                          const vehicle * const source_veh )
+                          const vehicle * const source_veh, 
+						  const monster * const source_mon)
```
Needs another space

Leland commented on src/creature.h:
```diff
@@ -544,6 +544,10 @@ class Creature
+		// Turn since creature was milked for last time
+		int turn_next_milking = 0;
+
+
```
Kill this extra line - just use one

Zilenan91 commented:
Wait so it makes unpasteurized milk? Wouldn't that be super unsafe to drink unless the player pasteurizes it?

AlecWhite commented:
It isn't that unsafe. At most I would give it a 5% chance of getting food poisoning.
Think that milk is full of lactobacillus that makes the milk harder for other bacteria to grow, unless it has been infected directly with ecoli or salmonella.
Raw milk doesn't "spoil" thanks to that bacteria, it just turns into clabbered milk.

BevapDin commented on src/creature.h:
```diff
@@ -544,6 +544,9 @@ class Creature
         virtual bool is_symbol_highlighted() const;
 
+		// Turn since creature was milked for last time
+		int turn_next_milking = 0;
```
The comment and the member name seem to disagree: is this the time of the last (most recent) milking (always in the past) or is it the time of the next milking (in the future)?

BevapDin replied:
As this only every applies to milked monsters (which would be *very* few), it might be better to make this an effect. Something like "was_milked" that times out after a day.

Also: the member is currently not saved at all. Saving and loading a game will just reset this to 0. Effects are automatically saved.

AlecWhite replied:
I don't know if it would be better to make some other animal recharge its milk faster or just make it a once per day thing. Or even if make it configurable on the json.

AlecWhite replied:
I just realize that I need to add some variable for milk charges left, in case you don't milk all the milk out of the udders.
I'm thinking of milking animals like toilets with milk instead of water.

BevapDin commented on src/game.cpp:
```diff
@@ -7686,7 +7686,13 @@ void game::examine( const tripoint &examp )
 
-        npc *np = dynamic_cast<npc*>( c );
+        if( mon != nullptr && mon->has_flag( MF_MILKABLE ) ) {
+            monster *source_mon = dynamic_cast<monster *>( c );
```
Why are you casting `c` again instead of using `mon`?

BevapDin commented on src/game.cpp:
```diff
@@ -9792,7 +9799,13 @@ bool game::handle_liquid( item &liquid, item * const source, const int radius,
+	{
+	    menu.text = string_format( _( "What to do with the %s from the %s?" ), liquid_name.c_str(),
+	                               source_mon->disp_name().c_str() );
+	} else
```
Please read and follow doc/CODE_STYLE.md. This should be `} else {` all on one line.

BevapDin commented on src/game.h:
```diff
@@ -643,8 +644,9 @@ class game
-                            const tripoint *source_pos = nullptr,
-                            const vehicle *source_veh = nullptr );
+                    const tripoint *source_pos = nullptr,
+                    const vehicle *source_veh = nullptr,
```
Unrelated whitespace changes, why?

AlecWhite replied:
Re-run it through Astyle.

BevapDin commented on src/iexamine.cpp:
```diff
@@ -2780,6 +2782,18 @@ void iexamine::water_source(player &p, const tripoint &examp)
+    if( source_mon->turn_next_milking < calendar::turn ) {
+        item milk( "milk", 0, 4 );
+        g->handle_liquid( milk, nullptr, 0, nullptr, nullptr, source_mon );
+        add_msg( "You milk the cow." );
```
String needs translation (add `_()`).

BevapDin replied:
Have you tested what happens when the player has no container and has to cancel the liquid handling?

AlecWhite replied:
I don't know how to use `(add _())` but I think I may have fixed.

Didn't thought about that scenario regarding having no containers, thank you for pointing it out. 
I'm going to try to see how to make it so you can save un-used charges inside the cow.

AMurkin replied:
About _() : [CONTRIBUTING.md#translations](https://github.com/CleverRaven/Cataclysm-DDA/blob/master/CONTRIBUTING.md#translations)

AlecWhite replied:
Thanks, I was kinda getting dizzy searching through the source files.

BevapDin commented on src/iexamine.cpp:
```diff
@@ -2780,6 +2782,18 @@ void iexamine::water_source(player &p, const tripoint &examp)
+        add_msg( "You milk the cow." );
+        source_mon->turn_next_milking = calendar::turn.get_turn() + HOURS( 24 );
+    } else {
+        add_msg( "The cow's udders run dry" );
```
Needs translation.

kevingranade commented:
> Wait so it makes unpasteurized milk? Wouldn't that be super unsafe to drink unless the player pasteurizes it?

Fresh raw milk is totally safe, it just spoils faster.

Night-Pryanik commented:
> Fresh raw milk is totally safe, it just spoils faster.

Really? As far as I know, it's highly dangerous. 
https://www.fda.gov/Food/resourcesForYou/consumers/ucm079516.htm

Rivet-the-Zombie commented:
Those poor babies, drinking lethal un-pasteurized milk!

Raw milk is quite safe; as safe as milk, even. FDA regulations aren't always grounded in reality.

Mecares commented:
> Really? As far as I know, it's highly dangerous.

Well pasteurization is done since 1864, drinking milk has been done for thousands of years before that, it is unlikely to be very dangerous.

AlecWhite commented:
Another Travis error. 

How do you read/what to look for in Travis information?

AMurkin commented:
It's only clang and not your code. Maybe it's mine, about NULLs.

Night-Pryanik commented:
> FDA regulations aren't always grounded in reality.

Of course there are stupid guys working at FDA. 

FDA was just the first link at Google, and many other health organizations will confirm that.

Night-Pryanik commented:
> Well pasteurization is done since 1864, drinking milk has been done for thousands of years before that, it is unlikely to be very dangerous.

I recommend reading on statistics of health problems, life longevity and death rate from food infections prior to discovering that aforementioned bad things are caused by pathogens in food.

ChuckJKelly commented:
Unpasteurized milk can be unsafe and grow bacteria when it's stored or otherwise handled incorrectly. But drinking it fresh shouldn't be an issue. Otherwise babies would be getting infections more or less constantly. As Kevin mentioned, it would just spoil faster.

>FDA regulations aren't always grounded in reality.
This is borderline insane though. Stored and shipped milk better damn well be pasteurized unless you want an awful lot of sick people.

AlecWhite commented:
> It's only clang and not your code. Maybe it's mine, about NULLs.

I remember yesterday Kevin say something in the IRC about Travis and cloud-based continuous integration system and about something being tricky.

Leland commented:
You are, as an individual, *highly* unlikely to suffer illness from unpasteurized milk. And, also, were the general populace today to start drinking milk unpasteurized entirely, there would be a large rise in illnesses. Both of these things are true.

AlecWhite commented:
And can't manage to undid a commit, will keep trying commands and whatever help Google's offer me.

Night-Pryanik commented:
Just a couple of statements to check. If I'm wrong somewhere, please correct me. 

Cows can be hosts to pathogens. 
Pathogens can pass from cows to their milk. 
We can't guarantee (without proper tools and analysis) that raw cow's milk is free from pathogens. 
People drinking raw cow's milk can become infected. 
So, raw milk is at least potentially dangerous.

kevingranade commented:
These statements:
> Pathogens can pass from cows to their milk.
> People drinking raw cow's milk can become infected.

Are quite flawed.  They are both vacuously true.
While it is true that pathogens can (and do) exist in raw milk, this is also true of nearly every substance we encounter,  the important distinction is the concentration and details of pathogens,  not their mere existence. Pathogens in milk become dangerous when allowed to multiply, but when fresh,  the pathogen count in milk is going to be very low. 
This leads to the second statement, very nearly all food we ingest contains*potentially* dangerous pathogens,  but in practice,  our immune systems are extremely good at eliminating pathogens in ingested foods, so this danger is dealt with as a matter of course. 

In practice, fresh raw milk is extremely safe, and anti-raw milk laws are mostlya result of lobbying by large dairy conglomerates rather than a necessary safety measure.

Night-Pryanik commented:
> mostlya result of lobbying by large dairy conglomerates rather than a necessary safety measure.

Sounds like a conspiracy theory to me. Conglomerates large enough to affect the whole world? Because in my country we have similar anti-raw milk laws. 

I will never drink raw milk (and won't eat raw eggs, for example, for the same reason), because I trust the guys in white smocks in FDA, WHO and other serious organizations.

Night-Pryanik commented:
>but when fresh, the pathogen count in milk is going to be very low.

BTW, is there a link to statistics asserting that?

ChuckJKelly commented:
The issue is that while drinking raw milk from cows on your farm the day you milk the cow is fine, drinking raw milk that was milked from a cow on a farm hundreds of miles away last week and shipped and packaged on an industrial scale is not.

The only thing I could find about pasteurized milk not being necessary for industrial scale is people also claiming that the government pasteurizes milk to remove vital nutrition from it so that the people remain weak and unable to rebel. If you believe that, I have a bridge you might be interested in buying...

DangerNoodle commented:
> Sounds like a conspiracy theory to me. Conglomerates large enough to affect the whole world? Because in my country we have similar anti-raw milk laws.

Just because the risk is low does not mean that the risk is acceptably low, or at least an organization might have a different standard for acceptably low compared to a rural farmer. If one person gets sick from drinking fresh milk, this creates a problem from a litigation perspective, even more so if the packaging does not adequately warn the consumer of this risk.

In addition, as brought up earlier, pasteurization is also important for storing milk long enough to safely transport it.

AlecWhite commented:
Does `void Creature::load( JsonObject &jsin )` and  `void Creature::store( JsonOut &jsout )` in savegame_json.cpp load and save respectively the information about the monsters in those weird JSON functions?

I'm trying to figure out how the whole saving monster and it looks like it uses some JSON funtions to save and read it.

Any help regarding about how the game saves and store data relating to creatures would be appreciated.

kevingranade commented:
It's designed to be transparent, if you apply an effect to the monster, it
will be saved and loaded automatically.
You should not need to change the way monsters are saved for this feature,
for example, you could apply a "has been milked" effect with a duration of
24h, and when a player milks the creature it uses the time remaining on
that effect to determine milk yield, and then re applies the effect.

AlecWhite commented:
@kevingranade I was thinking of adding a variable that saves when will the creature generate more milk and another one that saves how much milk has it left. 
As BevapDin pointed out, right now if you milk the cow and don't use all of the 4 charges they are lost. Also when you milk a cow save and quit and reload, the cow has milk again. 

So basically I'm trying to save 2 variables, charges of milk left and when will the cow(or creature) regenerate its milk stores.

Qrox commented:
What about increasing the effect's intensity by one, capped at some value, when its duration times out? And when the player milks the animal, decrease the effect's intensity accordingly.

kevingranade commented:
Extra variables are unnecessary, if you want to do something more sophisticated about partial milking you can e.g. prorate the effect duration you set in the "recently milked" effect.  We are not going to add variables to every monster to support functions that a tiny fraction of them use.

AlecWhite commented:
@kevingranade If a cow has 4 charges of milk, you milk 1 time, it could set up a "threequarters_milked" flag with a duration of 24 hours, but then if you milk it again 4 hours later, how can you track the time to add a flag with the correct time for the cow to regenerate the milk?

Night-Pryanik commented on data/json/effects.json:
```diff
@@ -1648,5 +1648,30 @@
+        "type": "effect_type",
+        "id": "half_milked",
+        "name": ["Half milked"],
+        "desc": ["This creature is milked to threequarters of its milk."]
```
Half, not three quarters.

Night-Pryanik commented on src/activity_handlers.cpp:
```diff
@@ -689,9 +698,20 @@ void activity_handlers::fill_liquid_do_turn( player_activity *act_, player *p )
         case LST_MAP_ITEM:
-            if( static_cast<size_t>( act.values.at( 1 ) ) >= source_stack.size() ) {
-                throw std::runtime_error( "could not find source item on ground for liquid transfer" );
+            if (static_cast<size_t>(act.values.at(1)) >= source_stack.size()) {
```
Why this change? It was correct styled.

AlecWhite replied:
I must mistakenly edited it, sorry.

kevingranade commented:
Wtf? Use the duration, having a bunch of different effects for this is going to be fragile and complicated.
In case it's not clear, if you fully drain the milk the cow would end up with a 24h duration effect, if you wait 6 hours, the remaining duration would be 18h , and you can calculate the available milk like ` (max_duration - current_duration) / max_duration`

AlecWhite commented:
>In case it's not clear, if you fully drain the milk the cow would end up with a 24h duration effect, if you wait 6 hours, the remaining duration would be 18h , and you can calculate the available milk like (max_duration - current_duration) / max_duration

I don't know if I fully understand it. Do you mean making an if to check the duration time in chunks of 6 hours, and adding it to max of 24, in which case another 6 hours should pass for the next milking?
Because if you mean that, then that would indeed only need one effect and all that jazz I made with the effects would not be needed. 
Gonna try it, althou I still have to find a way to eliminate the error I get due to not knowing how to pass the source_mon to the fill_liquid_do_turn.

BorkBorkGoesTheCode commented:
I like this idea. Its time has come.

Re: Pasteurization: Both views of milk safety seem to be half correct. Raw milk does seem to be safe immediately after milking, but it can serve as a growth medium for microbes if not stored in a cold, sanitary environment, leading to spoilage and food safety issues due to microbial action.

kevingranade commented:
> I don't know if I fully understand it. Do you mean making an if to check the duration time in chunks of 6 hours, 

No, I mean at any point in time you can calculate the milk in the creature's udders like:
```
available_milk = max_milk_capacity * (max_duration - current_duration) / max_duration;
```
and after a partial milking you can reverse the process:
```
current_duration = max_duration * (available_milk / max_milk_capacity);
```

DangerNoodle commented:
> Re: Pasteurization: Both views of milk safety seem to be half correct. Raw milk does seem to be safe immediately after milking, but it can serve as a growth medium for microbes if not stored in a cold, sanitary environment, leading to spoilage and food safety issues due to microbial action.

Related is that storage at the store, specially on the shelf, is not even remotely an adequate environment. Stocking takes time, and boxes of frozen goods can often end up sitting outside the displays for half an hour or longer, having already been hauled out on pallets even longer still.

And that is merely on the employee side, the consumer has no way of verifying that another customer did not pick up frozen goods, cart them around for prolonged periods of time, then change their mind and put them back, or worse leave them on regular shelves. And following that, the risk of a careless employee finding such a thing and deciding it will be fine to put back in the cold section.

Night-Pryanik commented:
> Raw milk does seem to be safe immediately after milking

Once again, what is the source of this information? 
Infected animal produce infected milk, isn't it? So raw milk is safe only if animal is healthy. And animal is healthy only if serious actions are taken by veterinarians. Which are sparse in post-cataclysm world, obviously.

kevingranade commented:
Please take the raw milk discussion elsewhere, perhaps smf.cataclysmdda.com, it's distracting here and any further comments about it will be deleted.

AlecWhite commented:
I think I managed to revert again without breaking anything important.

Night-Pryanik commented on data/json/effects.json:
```diff
@@ -1648,5 +1648,11 @@
+        "type": "effect_type",
+        "id": "milked",
+        "name": [ "Milked" ],
+        "desc": [ "The creature has been partically milked." ],
```
Partially?

BevapDin commented on src/game.cpp:
```diff
@@ -7685,8 +7685,12 @@ void game::examine( const tripoint &examp )
             }
         }
+        if( mon != nullptr && mon->has_flag( MF_MILKABLE ) ) {
+            add_msg( m_info, _( "This is a milkable cow." ) );
```
Actually it's a milkable monster, nothing here checks that it's a cow. You might want to add the monster type name instead of "cow".

AlecWhite replied:
I left it open on purpose to let people mod more animals, like alpacas, goats, and all kinds of mammals.

BevapDin replied:
But here it *explicitly* says "cow" and there is now way to change this. Even if I would add the flag to rabbits, it would say "This is a milkable cow."

AlecWhite replied:
Right, my bad. I forgot to put the thing that reads whatever animal you are looking.

AlecWhite commented on src/activity_handlers.cpp:
```diff
@@ -689,9 +700,19 @@ void activity_handlers::fill_liquid_do_turn( player_activity *act_, player *p )
+        case LST_MONSTER:
+        Creature *c = g->critter_at(source_pos);
+        source_mon = dynamic_cast<monster *>(c);
+        if( source_mon == nullptr ){
```
Could someone please help me with how the fill_liquid_do_turn function helps in activity_handlers.cpp?

Learned how to use breakpoints and looked at the c object and source_mon, both of them had the correct monster and location of it. But for some reason the throw at LST_MONSTER case is triggered even if source_mon isn't a null pointer.

AlecWhite replied:
With knocktwice's suggestion I realize that it is problem of the algorithm to set the effects. Working on it.

BevapDin commented on src/game.cpp:
```diff
@@ -9792,17 +9804,24 @@ bool game::handle_liquid( item &liquid, item * const source, const int radius,
-            // consume_item already consumes moves.
-            u.consume_item( liquid );
-        } );
+    // Small check to avoid players sucking milk directly from the animal.
```
Why? Why can't the PC drink directly from the animal?

AlecWhite replied:
I asked if the player should be able to suck milk directly from the udders, and the general consensus on the IRC channel was to disabled it.

BevapDin commented on src/iexamine.cpp:
```diff
@@ -2780,6 +2783,29 @@ void iexamine::water_source(player &p, const tripoint &examp)
+    item milk( "milk", 0, 1 );
+
+    // Max duration set here to avoid including effect.h
+    int max_dur = 14400;
```
What's that magic number supposed to be? Now I have to look it up and I have to calculate this myself. Why not use `HOURS(24)` or similar?

BevapDin commented on src/iexamine.cpp:
```diff
@@ -2780,6 +2783,29 @@ void iexamine::water_source(player &p, const tripoint &examp)
+    if( !source_mon->has_effect( effect_milked ) ) {
+        g->handle_liquid( milk, nullptr, 0, nullptr, nullptr, source_mon );
+        add_msg( _( "You milk the %s." ), source_mon->disp_name().c_str() );
+    } else if( source_mon->has_effect( effect_milked ) &&
```
This `has_effect` check is redundant. The branch will only be taken when the monster has the effect, otherwise the first `if` condition already yields `true`.

BevapDin commented on src/game.cpp:
```diff
@@ -9792,17 +9804,24 @@ bool game::handle_liquid( item &liquid, item * const source, const int radius,
-            u.consume_item( liquid );
-        } );
+    // Small check to avoid players sucking milk directly from the animal.
+    if( !source_mon == true ){
```
The ` == true` is redundant and therefor not needed.

BevapDin commented on src/activity_handlers.cpp:
```diff
@@ -689,9 +700,19 @@ void activity_handlers::fill_liquid_do_turn( player_activity *act_, player *p )
+        liquid.deserialize( act.str_values.at( 0 ) );
+        break;
+
             on_ground = source_stack.begin();
```
These lines look like they belong to the previous case.

AlecWhite replied:
I thought I commented here but I don't know where that comment went, guess I don't know how github works yet.

Regarding the lines, I don't really know how the whole function works, including why it serialized/deserialized things, I have the vague idea that they have to do with checking the json for the liquid, but not much really.

Also already asked help in a review about a part of the code in that function. But then I don't know if that's how you ask for help or how you point out to what you have a problem with.

BevapDin commented on src/activity_handlers.cpp:
```diff
@@ -758,6 +779,26 @@ void activity_handlers::fill_liquid_do_turn( player_activity *act_, player *p )
             break;
+        case LST_MONSTER:
+            Creature *c = g->critter_at(source_pos);
+            source_mon = dynamic_cast<monster *>(c);
```
These two lines are redundant, you have already set up (and checked) the content of `source_mon`. The pointer is still fine at this point and does not need to be recalculated.

AMurkin commented:
Dont you forget flags in mtype.h?

AlecWhite commented:
>Dont you forget flags in mtype.h?

I completely forgot about them, thanks for pointing it out.

AlecWhite commented:
I think this is ready. You get to milk the cows for 1 milk 3 times per 24 hours. 
And you can't suck milk directly from the cow's udders.

What do you guys think?

AMurkin commented:
I wanna milk, but can it be tested?
No creatures with flag MF_MILKABLE, or i missed it?

AlecWhite commented:
>I wanna milk, but can it be tested?
>No creatures with flag MF_MILKABLE, or i missed it?

Everything is working, and working as inteded. I think.

Only cows can be milkable. Althou it still needs a diary farm with a tamed/genetically engineered/bred version for "industrial" cows. As you will probably need the Animal Empathy to be able to get close to a cow without it attacking you or fleeing from you.

AMurkin commented:
What i've spotted:
1. [here you messed up two cases, and wrong indentation](https://github.com/AlecWhite/Cataclysm-DDA/blob/97ff9ff645b31247e6129741f5c15d21a6a2ca90/src/activity_handlers.cpp#L712-L728)
2. [odd space](https://github.com/AlecWhite/Cataclysm-DDA/blob/97ff9ff645b31247e6129741f5c15d21a6a2ca90/src/activity_handlers.cpp#L815)
3. [odd empty line and tabs at next lines](https://github.com/AlecWhite/Cataclysm-DDA/blob/97ff9ff645b31247e6129741f5c15d21a6a2ca90/src/game.cpp#L9807)
4. [no spaces](https://github.com/AlecWhite/Cataclysm-DDA/blob/97ff9ff645b31247e6129741f5c15d21a6a2ca90/src/iexamine.cpp#L56)
5. [tabs](https://github.com/AlecWhite/Cataclysm-DDA/blob/97ff9ff645b31247e6129741f5c15d21a6a2ca90/data/json/effects.json#L1652-L1658)

AlecWhite commented:
Thanks a lot Murkin.

Now to see why gorgon and travis doesn't like these commits.

AMurkin commented:
Hope it will be fixed by #21134

Leland commented:
Should you be able to milk hostile creatures?

I ask this because cows are apparently completely hostile and will attack you on sight. That should probably be changed as well.

AlecWhite commented:
>Should you be able to milk hostile creatures?

>I ask this because cows are apparently completely hostile and will attack you on sight. That should probably be changed as well.

If this PR gets merged I have thought modifying the dairy farm cows to some "genetically enhanced" or "tamed" variant that would be near brain dead/suicidal. Making you able to milk it but the cow won't run away from danger.

kevingranade commented:
> Should you be able to milk hostile creatures?

Ideally (not something I'm asking for in this PR or even soon) you'd restrain the animal somehow before milking it, and hostility would be irrelevant.  AFAIK cows are normally restrained when being milked, and you'd only milk a VERY tame cow without restraint.

Short term yes, limit it to "tame" animals, and have them stand still while being milked.

DangerNoodle commented:
The simplest expedient would be to have cows currently be passive, and just assume they are well tamed or genetically altered. Any actual implementation of more involved interaction with animals would warrant a separate pull request, as part of a general fleshing out of interacting with domestic animals.

Alternatively, the player might be allowed some way to attempt to get the animal used to their presence, or implement a somewhat game-ish method of taming via giving them food. As it stands, dogs can already be acclimated to the survivor with 100 percent reliability, regardless of how stray or feral they might be, how acclimated they are to unfamiliar people, et cetera.

Leland commented:
I think cows should be pacified. 

@AlecWhite I realize pacifying cows may seem out of scope for this PR but I think you'd agree that there should be at least one viable, milkable animal that comes out of this.

Also, cows are [in the `GROUP_SAFE` monster group](https://github.com/CleverRaven/Cataclysm-DDA/blob/bfc977cce2d070be2ff0dd36433ca9e34e71be88/data/json/monstergroups.json#L1232) and yet will attack players on sight. I have a feeling that the aggression setting for cows was unintentionally set too high, but they're spawned so rarely that it's never really been an issue. They only spawn at apple orchards, ponds, and farms; locations which range from pretty rare to perhaps never spawned at all.

Implementation details are up to you, but if you decide to go with pacifying them, I'd recommend [setting `mon_cow`'s aggression](https://github.com/CleverRaven/Cataclysm-DDA/blob/b3d1229085879e925d70c70ccf2c144bd7154c58/data/json/monsters.json#L1165) to -99, and consider removing the second sentence from it's description (describing the male cows' "vicious streak").

DangerNoodle commented:
Definitely lower cow aggression in this pull request for now, and other solutions can be devised in the future.

Coolthulhu commented on src/activity_handlers.cpp:
```diff
@@ -695,6 +707,15 @@ void activity_handlers::fill_liquid_do_turn( player_activity *act_, player *p )
+            Creature *c = g->critter_at(source_pos);
+            source_mon = dynamic_cast<monster *>(c);
+            if (source_mon == nullptr) {
+                throw std::runtime_error("could not find source creature for liquid transfer");
```
Last time I checked, we don't use `throw` for non-critical things. Just `debugmsg` and try to restore to some semi-sane state (just cancel everything).

Coolthulhu commented on src/game.cpp:
```diff
@@ -9792,17 +9804,22 @@ bool game::handle_liquid( item &liquid, item * const source, const int radius,
-        } );
+    // Small check to avoid players sucking milk directly from the animal.
+    if( !source_mon ) {
+        if( u.can_consume( liquid ) ) {
```
Would be cleaner as `if( !source_mon && u.can_consume( liquid ) )`

DangerNoodle commented:
Welcome back, @Coolthulhu. I was beginning to get worried about you.

AlecWhite commented:
And time to see again how to revert a commit.

AMurkin commented on src/activity_handlers.cpp:
```diff
@@ -712,7 +712,7 @@ void activity_handlers::fill_liquid_do_turn( player_activity *act_, player *p )
             source_mon = dynamic_cast<monster *>(c);
             if (source_mon == nullptr) {
-                throw std::runtime_error("could not find source creature for liquid transfer");
+                debugmsg("could not find source creature for liquid transfer");
```
no return - may and will cause error later

AMurkin commented on src/game.cpp:
```diff
@@ -9811,16 +9811,14 @@ bool game::handle_liquid( item &liquid, item * const source, const int radius,
-    if( !source_mon ) {
-        if( u.can_consume( liquid ) ) {
+
+        if( u.can_consume( liquid ) && !source_mon ) {
```
wrong indentation

AlecWhite commented:
Poking @BevapDin for requested changes review please.

BrettDong commented on src/iexamine.cpp:
```diff
@@ -2780,6 +2783,19 @@ void iexamine::water_source(player &p, const tripoint &examp)
+        g->handle_liquid( milk, nullptr, 0, nullptr, nullptr, source_mon );
+        add_msg( _( "You milk the %s." ), source_mon->disp_name().c_str() );
+    } else {
+        add_msg( _( "The %s's udders run dry" ), source_mon->disp_name().c_str() );
```
Should add a comma.

AlecWhite replied:
Wouldn't it be a dot?

BrettDong replied:
Sorry, I wanted to say a period...

BrettDong replied:
Sorry, I wanted to say a period...

BrettDong commented on src/activity_handlers.cpp:
```diff
@@ -758,8 +783,22 @@ void activity_handlers::fill_liquid_do_turn( player_activity *act_, player *p )
             // nothing, the liquid source is infinite
             break;
+        case LST_MONSTER:
+        if( source_mon->has_effect( effect_milked ) ) {
```
Would be better if this block of code is indented.

Coolthulhu commented on src/game.cpp:
```diff
@@ -9779,7 +9785,13 @@ bool game::handle_liquid( item &liquid, item * const source, const int radius,
             return true;
-        } else {
+        }
+        else if( source_mon != nullptr ){
```
Styling:
``
} else if( x ) {
```

Coolthulhu commented on src/game.cpp:
```diff
@@ -9792,12 +9804,15 @@ bool game::handle_liquid( item &liquid, item * const source, const int radius,
+    } else if( source_mon != nullptr ) {
+	    menu.text = string_format( _( "What to do with the %s from the %s?" ), liquid_name.c_str(),
+	                               source_mon->disp_name().c_str() );
+	} else {
```
Tab, check all the changed files if you don't have any other missed tabs. Most text editors should support regular search detecting tabs.

AlecWhite replied:
Didn't know you could search for tabs, neat.

Coolthulhu commented on src/iexamine.cpp:
```diff
@@ -2780,6 +2783,19 @@ void iexamine::water_source(player &p, const tripoint &examp)
     g->handle_liquid( water, nullptr, 0, &examp );
 }
 
+void iexamine::milk_source( monster *source_mon )
```
Use references instead of pointers where possible, unless you expect the monster to be `null` at times.

The placement of this function in `iexamine` is not exactly correct. It's not a huge error (just a matter of convention) and I'd allow it, but from what I know we're in feature freeze right now so you have time to fix it.
The proper place for it would be a new pair of files, something like `monexamine.h/.cpp` or `monsterexamine.h/.cpp`.
`iexamine` is not the proper place because it handles examination of specific class of objects - terrain and furniture. Traps for example are excluded.
The proper place for it would be either a new pair of files

AlecWhite replied:
>Use references instead of pointers where possible, unless you expect the monster to be null at times.

I'm very noob with C++ so I have absolutely not idea of what's the difference between using pointer and reference in this situation, so I just copied how the other situations with vehicles, item source and map worked. Would you mind explaining it please? @Coolthulhu

>It's not a huge error (just a matter of convention)

I don't understand what you mean by "convention", is there any info regarding conventions? as to not commit this mistake again.

> I'd allow it, but from what I know we're in feature freeze right now so you have time to fix it.

Then should I make a new pair of files or leave it as it is?

>The proper place for it would be either a new pair of files

I feel like you wanted to say something more(by that either) and got cut short.

Coolthulhu replied:
> I'm very noob with C++ so I have absolutely not idea of what's the difference between using pointer and reference in this situation

If you don't expect the argument to be `null`, use a reference.
Pointers can be `null` (you need to check if they aren't or document that they can't be), can point to different objects (that is, you can "reseat" a pointer) and use a different syntax that accounts for those two properties.
For someone new to C++, it is safe to use references pretty much everywhere you can and only use pointers if you can't use a reference or you have a pointer in the block above and are just using that variable.

> I don't understand what you mean by "convention", is there any info regarding conventions? 

I mean just as good practice, not codified convention.
In general, objects should be grouped by their roles/types/etc. and not share them if that is not necessary. In this case, `iexamine` handles all functions with specific set of arguments - this is useful, because we can then bind those functions to specific objects and use it like:
```
terrain_type.on_examine( who, where );
```

Your examine function is an examine function, but not "examine point by person" function, hence it would fit better somewhere else.

> I feel like you wanted to say something more(by that either) and got cut short.

Oops
I wanted to say that you can either make new files or have it all in `game.cpp`. Putting new functions to `game.cpp` is a bad practice at this point (this file is huge), but if you want to limit the changes in this PR, this would be the easiest way to go.
If you don't care about limiting number of changes, create the two new files - it will be cleaner and will be useful later if you want to add more Harvest Cataclysm functions, like shearing wool, giving animals meds and so on.

Coolthulhu commented on src/iexamine.cpp:
```diff
@@ -2780,6 +2783,19 @@ void iexamine::water_source(player &p, const tripoint &examp)
+void iexamine::milk_source( monster *source_mon )
+{
+    item milk( "milk", 0, 1 );
+    if( !source_mon->has_effect( effect_milked ) ) {
```
Why does this case print no message?
If it is supposed to print it, you could just skip this case and handle it by one below: if a creature has no effect, `get_effect_dur` will return 0 anyway.

AlecWhite replied:
Decided to combine those two if in one as I haven't realized that I could do that instead of making 2 if for each particular case.

Coolthulhu commented on src/game.h:
```diff
@@ -644,7 +645,8 @@ class game
                             const tripoint *source_pos = nullptr,
-                            const vehicle *source_veh = nullptr );
+                            const vehicle *source_veh = nullptr,
+                            const monster *source_mon = nullptr );
```
This function is growing awfully huge and multi-purpose.
5 optional arguments, 4 of which default to `null` is a good sign that this will need a refactor soon.
That extra argument is pushing it, but could be tolerated for now.

Coolthulhu commented on src/game.cpp:
```diff
@@ -7685,8 +7685,12 @@ void game::examine( const tripoint &examp )
         }
+        if( mon != nullptr && mon->has_flag( MF_MILKABLE ) ) {
+            add_msg( m_info, _( "This is a milkable %s." ), mon->name().c_str());
+            iexamine::milk_source( mon );
```
Another minor reason why `milk_source` shouldn't be in `iexamine`: you want to return non-`void` here, because you want the examination to stop if player chooses to milk.
With your current version, you could examine a cow that stands on a trap (say, a mine), start milking the cow, then immediately also try to defuse the mine and blow the cow up.

Would be better if you integrated it with `pet_menu` above. Changed `pet_menu` to `monster_menu`, allowed pet effects only if the monster is a pet, but allowed milking if it is milkable. That way pet cows wouldn't require two menus, and I assume later on only pet cows will be milkable anyway.

AlecWhite replied:
Wouldn't that require to move all the pet menu and all that to monexamine.h/.cpp or monsterexamine.h/.cpp? I mean, I could try to do it, but I really don't know how much help I would need to make it work.

Coolthulhu replied:
> Wouldn't that require to move all the pet menu and all that to monexamine.h/.cpp or monsterexamine.h/.cpp?

Not require, though it could include that.
You could leave the `pet_menu` in `game.cpp`, just have it refer to something in `monexamine` (one menu entry).

> Also as far as I know cows are "tamed" by way of manually selecting the most passive ones rather than more actively training(like a dog).

It should be safe to assume that all the cows that survived do fear anything that looks like a human.

AlecWhite replied:
>It should be safe to assume that all the cows that survived do fear anything that looks like a human.

Actually, researched more on cow taming and things like this pops up a lot https://www.youtube.com/watch?v=vXw620-Tc9Q.

Is there any way to get an item's ID? I can't find any method that does it on item.h, I may have to make one myself.

Zireael07 commented:
Why close?

AlecWhite commented:
Closed in favour of https://github.com/CleverRaven/Cataclysm-DDA/pull/21272. As I would have need to reset everything and all that.
